hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -351,30 +351,80 @@ async def attempts(self):
 
 
 class BatchSubmissionInfo:
-    def __init__(self, used_fast_create: Optional[bool] = None, used_fast_update: Optional[Dict[int, bool]] = None):
-        self.used_fast_create = used_fast_create
-        self.used_fast_update = used_fast_update or {}
+    def __init__(self, used_fast_path: Optional[bool] = None):
+        self.used_fast_path = used_fast_path
+
+
+class BatchNotSubmittedError(Exception):
+    pass
+
+
+class BatchAlreadyCreatedError(Exception):
+    pass
+
+
+def submitted_batch_only(fun):
+    @functools.wraps(fun)
+    async def wrapped(*args, **kwargs):
+        batch = args[0]
+        if not batch.submitted:
+            raise BatchNotSubmittedError
+        return await fun(*args, **kwargs)
```

Comment:
I really like how you've structured this to be explicit about what operations are valid in each state that a `Batch` object can be in. Unfortunately doing this as a decorator means you'll need different decorators for sync and async `fun`s. It also erases all type information from the signature of the method (I tried for a while to type this and was struggling). Maybe these can just be private methods in the class and instead of a decorator other methods should start with `self._raise_if_unsubmitted()`? It would still be the same number of lines, albeit perhaps slightly noisier.